mobile_app_connector/models/account_invoice.py

```diff
@@ -33,92 +51,184 @@ def mobile_post_invoice(self, json_data, **parameters):
         self.gift_treated = json.get('LastInsertedDonationId') or \
             json.get('LastInsertedGiftId')
         self.source = json.get('source', 'iOS')
 
-    @staticmethod
-    def _get_gift_products(gift_types, env):
-        """
-        Maps the app fund names to our gift products
-        :param gift_types: list of product template ids sent by the app
-        :param env: odoo environment
-        :return: Recordset of product.product
-        """
-        product_obj = env['product.product']
-        products = product_obj
-        for gift_type in gift_types:
-            product = product_obj.search([
-                ('product_tmpl_id', '=', gift_type)], limit=1)
+        self.fund_donations = []
+        self.sponsorships_payments = []
+        self.child_gifts = []
+
+        fund_ids = json.get('appealtype', [])
+        fund_amounts = json.get('appealamount', [])
+        gift_products = json.get('gifttype', [])
+        gift_amounts = json.get('giftamount', [])
+        child_ids = json.get('need', [])
+
+        # Map fund donations to products
+        product_obj = env['product.product'].sudo()
+        for i, product_id in enumerate(fund_ids):
+            product = product_obj.search(
+                [('product_tmpl_id', '=', product_id)], limit=1)
+            if not product:
+                raise ValueError(_("The product %s was not found") % product_id)
+            self.fund_donations.append({
+                'product_id': product,
+                'amount': fund_amounts[i],
+            })
+
+        # Map gift_products (gifts and sponsorships payments)
+        for i, product_id in enumerate(gift_products):
+            product = product_obj.search(
+                [('product_tmpl_id', '=', product_id)], limit=1)
             if not product:
-                raise ValueError(
-                    _("The product %s was not found") % gift_type)
-            products += product
-        return products
+                raise ValueError(_("The product %s was not found") % product_id)
+
+            spn = env['recurring.contract'].search([
+                ('child_id', '=', child_ids[i]),
+                '|', ('correspondent_id', '=', self.partner_id),
+                ('partner_id', '=', self.partner_id),
+                ('state', 'not in', ['terminated', 'cancelled'])
+            ]).ensure_one()
+            p = {
+                'contract_id': spn,
+                'child_id': child_ids[i],
+                'product_id': product,
+                'amount': gift_amounts[i],
+            }
+            if product.product_tmpl_id == env.ref(
+                    'sponsorship_compassion.product_template_sponsorship'):
+                self.sponsorships_payments.append(p)
+            else:
+                self.child_gifts.append(p)
+
+    def is_multiple_months_payment(self):
+        # check whether the user is trying to pay multiple times for a single contract
+        spns = {}
```

Comment:
You could use `defaultdict` from python to simplify the code (https://docs.python.org/2/library/collections.html#collections.defaultdict)